src/util.cpp

```diff
@@ -817,6 +808,49 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
 #endif
 }
 
+FILE* AdviseSequential(FILE *file) {
+#if _POSIX_C_SOURCE >= 200112L
+    // Since this whole thing is advisory anyway, we can ignore any errors
+    // encountered up to and including the posix_fadvise call. However, we must
+    // rewind the file to the appropriate position if we've changed the seek
+    // offset.
+    if (file == nullptr)
+        return nullptr;
+    int fd = fileno(file);
+    if (fd == -1)
+        return file;
+    off_t start = lseek(fd, 0, SEEK_CUR);
+    if (start == -1)
+        return file;
+    off_t end = lseek(fd, 0, SEEK_END);
+    if (end != -1) {
+        posix_fadvise(fd, start, end - start, POSIX_FADV_WILLNEED);
+        posix_fadvise(fd, start, end - start, POSIX_FADV_SEQUENTIAL);
+    }
+    lseek(fd, start, SEEK_SET);
+#endif
+    return file;
+}
+
+int CloseAndDiscard(FILE *file) {
+#if _POSIX_C_SOURCE >= 200112L
+    // Ignore any errors up to and including the posix_fadvise call since it's
+    // advisory.
+    if (file != nullptr) {
+        off_t end;
+        int fd = fileno(file);
+        if (fd == -1)
+            goto close;
+        end = lseek(fd, 0, SEEK_END);
+        if (end == -1)
+            goto close;
+        posix_fadvise(fd, 0, end, POSIX_FADV_DONTNEED);
+    }
+#endif
+ close:
```

Comment:
Unused labels might be a warning in some compiler versions; move inside #if block?